### PR TITLE
busbus.web

### PR DESCRIPTION
**Pull request by ianweller**
Oh boy, a JSON web interface!

Everybody else in spaceboats is familiar with the web API as they've used it for their projects (see https://github.com/spaceboats/busbus-android and https://github.com/spaceboats/busbus-web-sample).

But for everybody else: this makes it so you can access data "the busbus way" without writing your code in Python, as long as you can make HTTP requests and parse JSON.

This branch notably does not contain documentation. That will be resolved before the end of the semester. Anything else is failure.

### CODE REVIEW

File: busbus/web.py [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+
+    def __init__(self, msg, error_code=500):
+        self.msg = msg
+        self.error_code = error_code
+
+
+class Engine(busbus.Engine):
+
+    def __init__(self, *args, **kwargs):
+        # perhaps fix this to use a decorator somehow?
```
Comment (Cybermite) [inline busbus/web.py]: Just a heads up, this line has "fix this"

Comment (ianweller) [inline busbus/web.py]: Making this work with a decorator is probably not within the scope I want for this pull request

File: busbus/web.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+            response['request']['action'] = action
+            if (entity, action) in self._entity_actions:
+                func, entity = self._entity_actions[(entity, action)]
+                try:
+                    response[entity] = unexpand_init(func(**kwargs), to_expand)
+                except APIError as exc:
+                    response['request']['status'] = 'error'
+                    response['error'] = exc.msg
+                    cherrypy.response.status = exc.error_code
+                return response
```
Comment (Cybermite) [inline busbus/web.py]: Since the "_limit" attribute is only checked in the next block, and this returns here, then any request with actions cannot utilize the "_limit" attribute.

i.e: stops/find?latitude=38.9717&longitude=-95.2353&distance=400&_limit=1
Gives me more than one stop.

File: busbus/web.py [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+        else:
+            entity_func = getattr(super(Engine, self), entity, None)
+            if entity_func is not None:
+                if '_limit' in kwargs:
+                    limit = int(kwargs.pop('_limit'))
+                    response['request']['limit'] = limit
+                result = getattr(super(Engine, self), entity).where(**kwargs)
```
Comment (Cybermite) [inline busbus/web.py]: For when a limit is given, is this doing unnecessary work? I mean if I want 1 arrival, is it still going to do the work for all the arrivals and then slice it down to one a couple lines down?

Comment (ianweller) [inline busbus/web.py]: Assuming the provider is using generators correctly, the values aren't generated until the unexpand_init() function looks at all of them. itertools.chain will stop the iterator after n items, so they never get generated if they're not asked for.

tl;dr: generator -> unexpand_init -> unexpand -> list

Comment (Cybermite) [inline busbus/web.py]: Okay, good. I figured it was a generator, but just had to be sure.

File: busbus/web.py (hunk 2 above)
Comment (ianweller) [inline busbus/web.py]: Fixed in f87d5bd

**Comment by Cybermite** [comment]
Good to go

**Comment by Cybermite** [comment]
Yea, good to go
